[flang][OpenMP] Convert repeatable clauses (except Map) in ClauseProc…

kparzysz opened the pull request:
…essor

Rename `findRepeatableClause` to `findRepeatableClause2`, and make the new `findRepeatableClause` operate on new `omp::Clause` objects.

Leave `Map` unchanged, because it will require more changes for it to work.

[Clause representation 3/6]

skatrak reviewed (commented):
This LGTM, I just have a question on a part of the patch that I wasn't able to fully understand.

kparzysz committed:
[flang][OpenMP] Implement flexible OpenMP clause representation

Introduce a set of generic classes (templates) that represent OpenMP
clauses in a language-agnostic manner. OpenMP clauses can contain
expressions and data objects and the exact representation of each
depends on the source language of the compiled program. To deal with
this, the templates depend on two type parameters:
- IdType: type that represent object's identity (in a way that
  satisfied OpenMP requirements), and
- ExprType: type that can represent numeric values, as well as
  data references (e.g. x.y[1].z[2]).

In addition to that, implement code instantiating these templates
from flang's AST.

This patch only introduces the new classes, they are not yet used
anywhere.

kparzysz committed:
[flang][OpenMP] Convert unique clauses in ClauseProcessor

Temporarily rename old clause list to `clauses2`, old clause iterator
to `ClauseIterator2`.
Change `findUniqueClause` to iterate over `omp::Clause` objects,
modify all handlers to operate on 'omp::clause::xyz` equivalents.

kparzysz committed:
[flang][OpenMP] Convert repeatable clauses (except Map) in ClauseProcessor

Rename `findRepeatableClause` to `findRepeatableClause2`, and make the
new `findRepeatableClause` operate on new `omp::Clause` objects.

Leave `Map` unchanged, because it will require more changes for it to
work.

kparzysz commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -2865,53 +2865,45 @@ bool ClauseProcessor::processDepend(
+            genDependKindAttr(firOpBuilder, clause);
+        dependTypeOperands.append(objects.size(), dependTypeOperand);
+
+        for (const omp::Object &object : objects) {
+          assert(object.dsg && "Expecting designator");
+
+          if (Fortran::evaluate::ExtractSubstring(*object.dsg)) {
+            TODO(converter.getCurrentLocation(),
+                 "substring not supported for task depend");
+          } else if (Fortran::evaluate::IsArrayElement(*object.dsg)) {
+            TODO(converter.getCurrentLocation(),
+                 "array sections not supported for task depend");
+          }
+
+          Fortran::semantics::Symbol *sym = object.sym;
           const mlir::Value variable = converter.getSymbolAddress(*sym);
           dependOperands.push_back(variable);
         }
```
The old code is below for reference.

First, the code created (unconditionally) a list of designators from the list of objects.  In my implementation, if an object was a designator, the designator would be present in `omp::Object`.  If the object is not a designator, the code would fail at runtime in both cases.

Then, a `DataRef` object is obtained from the designator, if present, but that's only to get the symbol out of it.  The other two cases (Substring and ArrayElement) are both TODO, which is what my code does as well.

```
        const std::list<Fortran::parser::Designator> &depVal =
            std::get<std::list<Fortran::parser::Designator>>(
                std::get<Fortran::parser::OmpDependClause::InOut>(
                    dependClause->v.u)
                    .t);
        mlir::omp::ClauseTaskDependAttr dependTypeOperand =
            genDependKindAttr(firOpBuilder, dependClause);
        dependTypeOperands.insert(dependTypeOperands.end(), depVal.size(),
                                  dependTypeOperand);
        for (const Fortran::parser::Designator &ompObject : depVal) {
          Fortran::semantics::Symbol *sym = nullptr;
          std::visit(
              Fortran::common::visitors{
                  [&](const Fortran::parser::DataRef &designator) {
                    if (const Fortran::parser::Name *name =
                            std::get_if<Fortran::parser::Name>(&designator.u)) {
                      sym = name->symbol;
                    } else if (std::get_if<Fortran::common::Indirection<
                                   Fortran::parser::ArrayElement>>(
                                   &designator.u)) {
                      TODO(converter.getCurrentLocation(),
                           "array sections not supported for task depend");
                    }
                  },
                  [&](const Fortran::parser::Substring &designator) {
                    TODO(converter.getCurrentLocation(),
                         "substring not supported for task depend");
                  }},
              (ompObject).u);
          const mlir::Value variable = converter.getSymbolAddress(*sym);
```

skatrak reviewed (approved):
Thank you for the clarification, LGTM.

tblah commented on flang/lib/Lower/OpenMP/ClauseProcessor.cpp:
```diff
@@ -87,50 +87,44 @@ getSimdModifier(const omp::clause::Schedule &clause) {
+    const auto &value =
+        std::get<omp::clause::Allocate::Modifier::Allocator>(modifier->u);
+    mlir::Value operand =
+        fir::getBase(converter.genExprValue(value.v, stmtCtx));
```
Why did you drop `Fortran::semantics::GetExpr`?

tblah commented on flang/lib/Lower/OpenMP/ClauseProcessor.cpp:
```diff
@@ -181,45 +172,41 @@ genDependKindAttr(fir::FirOpBuilder &firOpBuilder,
-  auto &expr = std::get<Fortran::parser::ScalarLogicalExpr>(ifClause->v.t);
   mlir::Value ifVal = fir::getBase(
-      converter.genExprValue(*Fortran::semantics::GetExpr(expr), stmtCtx));
+      converter.genExprValue(std::get<omp::SomeExpr>(clause.t), stmtCtx));
```
Here too

tblah commented on flang/lib/Lower/OpenMP/ClauseProcessor.cpp:
```diff
@@ -181,45 +172,41 @@ genDependKindAttr(fir::FirOpBuilder &firOpBuilder,
-    useDeviceSymbols.push_back(sym);
-  }
+  for (const omp::Object &object : objects)
+    useDeviceSymbols.push_back(object.id());
```
Why have you dropped the call to `getOmpObjectSymbol`?

kparzysz replied:
Because `object.id()` returns `Symbol*`.  Also, same reason as above.

kparzysz replied:
All things that were previously in `parser::Something` were have been translated into the new clauses: `OmpObject` is now `Symbol*` plus designator (`MaybeExpr`), other things are mostly `SomeExpr`.  The old code that computed these is no longer needed in places where it used to be.

tblah reviewed (approved):
Thanks for explaining. LGTM

kparzysz committed:
Merge branch 'main' into users/kparzysz/spr/b03-rep1

kparzysz committed:
Reduce indentation by collapsing "else {if}" into "else if"

kparzysz commented on flang/lib/Lower/OpenMP/Utils.cpp:
```diff
@@ -28,9 +29,27 @@ namespace Fortran {
+    assert(sym && "Expected Symbol");
+    if (mlir::Value variable = converter.getSymbolAddress(*sym)) {
+      operands.push_back(variable);
+    } else {
```
Done

kparzysz committed:
clang-format
